client/consensus-spin/secondary_leader.md

```diff
@@ -0,0 +1,186 @@
+    *   Ensure it handles wrapping and the case of a single authority gracefully (where there might be no secondary).
+
+2.  **Modify `check_header_slot_and_seal`:**
+    *   Inside the function, after getting `slot` and `seal`/`sig`:
+    *   Calculate `primary_author` using `slot_author`. Get its index if needed.
+    *   Calculate `secondary_author` using `get_secondary_author`.
+    *   Perform the verification check:
+        ```rust
+        let pre_hash = header.hash();
+        let primary_valid = P::verify(&sig, pre_hash.as_ref(), primary_author);
+
+        let secondary_valid = if let Some(sec_author) = secondary_author {
+            P::verify(&sig, pre_hash.as_ref(), sec_author)
+        } else {
+            false
+        };
+
+        if primary_valid || secondary_valid {
+            // Signature is valid under either the primary or secondary expected author
+            Ok((header, slot, seal))
+        } else {
+            // Signature is not valid under either expected author
+            Err(SealVerificationError::BadSignature)
+        }
+        ```
+    *   Make sure `slot_author` correctly handles session boundaries and returns the index along with the author ID if needed for the secondary calculation.
+
+3.  **Modify Proposer (`worker.rs` or similar):**
+    *   In the main loop checking slots:
+    *   When considering proposing for `slot`:
+        *   `let primary_author = slot_author(slot, ...);`
+        *   `let primary_index = authorities.iter().position(|a| a == primary_author).unwrap();`
+        *   `let secondary_author = get_secondary_author(primary_index, authorities);`
+        *   `let local_keys = keystore.keys(...); // Get local authority keys`
+        *   `let i_am_primary = local_keys.contains(primary_author);`
+        *   `let i_am_secondary = secondary_author.map_or(false, |sa| local_keys.contains(sa));`
+
+        *   If `i_am_primary`:
+            *   `claim_slot` and propose immediately.
+        *   Else if `i_am_secondary`:
+            *   Calculate `deadline = slot_start_time + T_secondary_wait`.
+            *   Wait until `deadline`.
+            *   Check if a block for `slot` has been imported/seen.
+            *   If not seen: `claim_slot` (using the *secondary* key) and propose.
+
+4.  **Runtime:**
+    *   Add logic (likely in pallet implementing `frame_support::traits::OnInitialize`) to:
+        *   Read the current slot from the digest.
+        *   Determine `P_primary` and `P_secondary`.
+        *   Determine the actual author `P_signer` (this might need an inherent data provider in consensus passing the signer's public key, or potentially reading it from the seal digest if possible/reliable).
+        *   If `P_signer == P_secondary`, report offence for `P_primary`. Keep track of consecutive offences if needed for the "N" threshold.
+
+
+
+### Decision Drivers (Optional)
+
+
+### Alternatives Considered
```

Comment:
Can we consider an alternative to `T_wait`, I mean with with the secondary leader always on and producing blocks which the rest of the network doesn't finalize while there are blocks from the primary?